File: src/grpc_client.cpp
#include <iostream>
#include <memory>
#include <string>

#include <grpcpp/grpcpp.h>

#include "greeter.grpc.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;
using greeter::Greeter;
using greeter::HelloReply;
using greeter::HelloRequest;

class GreeterClient {
public:
    GreeterClient(std::shared_ptr<Channel> channel) : stub_(Greeter::NewStub(channel)) {}

    // サーバーにリクエストを送信
    std::string SayHello(const std::string &user) {
        // リクエストを準備
        HelloRequest request;
        request.set_name(user);

        // レスポンス用のコンテナ
        HelloReply reply;

        // コンテキストは、RPCの設定を変更するために使用可能
        ClientContext context;

        // 実際のRPC呼び出し
        Status status = stub_->SayHello(&context, request, &reply);

        // 返り値とレスポンスをチェック
        if (status.ok()) {
            return reply.message();
        }
        std::cerr << "RPC failed: " << status.error_code() << ": " << status.error_message() << "\n";
        return "RPC failed";
    }

private:
    std::unique_ptr<Greeter::Stub> stub_;
};

int main(int argc, char *argv[]) {
    // サーバーアドレス
    std::string target_str = "localhost:50051";

    // コマンドライン引数からユーザー名を取得、なければ "World"
    std::string user = (argc > 1) ? argv[1] : "World";

    // チャネルを作成（認証なし）
    GreeterClient greeter(grpc::CreateChannel(target_str, grpc::InsecureChannelCredentials()));

    std::cout << "Sending request to " << target_str << " with name: " << user << "\n";

    // RPCを呼び出し
    std::string reply = greeter.SayHello(user);
    std::cout << "Greeter received: " << reply << "\n";

    return 0;
}

File: python/grpc_client.py
#!/usr/bin/env python3
"""gRPC client example in Python."""

import sys
import grpc
import greeter_pb2
import greeter_pb2_grpc


def run():
    """Run the gRPC client."""
    # サーバーアドレス
    server_address = 'localhost:50051'

    # コマンドライン引数からユーザー名を取得、なければ "World"
    name = sys.argv[1] if len(sys.argv) > 1 else "World"

    # チャネルを作成（認証なし）
    with grpc.insecure_channel(server_address) as channel:
        stub = greeter_pb2_grpc.GreeterStub(channel)

        print(f"Sending request to {server_address} with name: {name}")

        # RPCを呼び出し
        response = stub.SayHello(greeter_pb2.HelloRequest(name=name))

        print(f"Greeter received: {response.message}")


if __name__ == '__main__':
    run()
